### ISSUE TITLE
Add `static_c_args` to library()

### ISSUE THREAD

**Issue by xclaesse** [mesonbuild/meson#3304]
It is common to pass different cflags depending if it's a static or shared library, for example -DSTATIC_COMPILATION.

I suggest adding `static_c_args` that will be appended to c_args if it's a static build. If static_c_args is not empty and `default_library=='both'` then it will have to build twice instead of extracting objects.

**Comment by sarum9in** [comment]
I believe static makes sense for multiple languages, e.g. C++
So if we are to implement this, we need to support more generic form.
Do we also need this for `pch`?
```text
<languagename>_pch precompiled header file to use for the given language
<languagename>_args compiler flags to use for the given language; eg: cpp_args for C++
++ static_<languagename>_args
```

**Comment by jpakkane** [comment]
Fortran, objc, Rust, D, and everything that compiles into native code. It could also be argued that you should make it possible to add args to shared libs and exes as well. Possibly also based on the build type. This gives `#languages x #targettypes #buildtypes` new kwarg parameters. This is not a feasible solution, since we try to keep the number of keyword arguments as low as possible.

**Comment by xclaesse** [comment]
Well, for static_c_args there are valid real world use-case: glib and gstreamer do have a -DSTATIC_COMPILATION to define dllexport on windows which is different for static/shared.

We could have static_args : {'c' : '-DSTUFF'}, but I'm not sure the interpreter supports dict, does it?

For shared_args I have no use case for it, but if there are, I don't think it's a big deal to add it later.

**Comment by xclaesse** [comment]
Just got an idea to make it scalable: Add a new type CompileArgs. Something like this:

args = compile_args(static_args : '-DSTATIC', shared_args : '-DSHARED')
lib = library(..., c_args : ['-DFOO', args])

Making existing <lang>_args accept a new generic type makes it extensible in the future. What do you think?

**Comment by jpakkane** [comment]
Or perhaps mirroring how `configure_file` works:

```meson
cargs = compile_args()
cargs.add_args('-foo', language: 'c', targettype : 'static_library')
library(.., extra_args : cargs)
```

?

**Comment by xclaesse** [comment]
Actually I like that idea of having an holder for all the build configs that can be passed to many library() executable() calls. It can replace add_global_arguments() for args that needs to be on most but not all targets.

It could also be extended to more than compiler args later. For example why not args.set_install_dir('/myprefix/lib', target_type: 'library'). All your settings could be in one place and you pass that object down to all your build targets. Anyway, that's for later if we see the need. For now the real use-case for me is static c flags.

**Comment by nirbheek** [comment]
I wonder if people need to pass arbitrary custom `<lang>_args` when doing a static compilation, or only defines. If the latter, the implementation is much simpler because it applies to C/C++/ObjC/ObjCpp at once.

**Comment by NickeZ** [comment]
I would argue that special defines for building dynamic libs is often the case on windows.

Sometimes it is even necessary to have special defines only when building with the intention of linking dynamically. Real world examples: hdf5 has to set `H5_BUILT_AS_DYNAMIC_LIB` and boost has to set `BOOST_ALL_DYN_LINK`.


```cpp
#ifdef _WIN32
#  ifdef MODULE_API_EXPORTS
#    define MODULE_API __declspec(dllexport)
#  else
#    define MODULE_API __declspec(dllimport)
#  endif
#else
#  define MODULE_API
#endif
```

Source: https://stackoverflow.com/a/19547601/3804521

One could also argue that `-fvisibility=hidden` is a good practice using gcc/clang and then the above macro becomes something like this:

```cpp
#if defined _WIN32 || defined __CYGWIN__
  #ifdef BUILDING_DLL
    #ifdef __GNUC__
      #define DLL_PUBLIC __attribute__ ((dllexport))
    #else
      #define DLL_PUBLIC __declspec(dllexport) // Note: actually gcc seems to also supports this syntax.
    #endif
  #else
    #ifdef __GNUC__
      #define DLL_PUBLIC __attribute__ ((dllimport))
    #else
      #define DLL_PUBLIC __declspec(dllimport) // Note: actually gcc seems to also supports this syntax.
    #endif
  #endif
  #define DLL_LOCAL
#else
  #if __GNUC__ >= 4
    #define DLL_PUBLIC __attribute__ ((visibility ("default")))
    #define DLL_LOCAL  __attribute__ ((visibility ("hidden")))
  #else
    #define DLL_PUBLIC
    #define DLL_LOCAL
  #endif
#endif
```

Source: https://gcc.gnu.org/wiki/Visibility

In both cases either `MODULE_API_EXPORTS` or `BUILDING_DLL` has to be set while building the dynamic version of a library on windows.

**Comment by xclaesse** [comment]
@nirbheek FWIW, my only use-case is glib/gstreamer's `-DSTATIC_COMPILATION`. IMHO it's a big enough use-case in itself to justify having a solution just for that, and that's why my initial PR wasn't a generic solution. But @jpakkane wanted to have something more generic.

**Comment by nirbheek** [comment]
> It was not immediately obvious to me that -DH5_BUILT_AS_DYNAMIC_LIB would be passed as /DH5_BUILT_AS_DYNAMIC_LIB on windows when I started using meson

fwiw, `/D` and `-D` are identical for `cl.exe`, and we pass that option unchanged. It's the same for all `/FOO` options passed to `cl.exe`, `link.exe`, and `lib.exe`. You're right, we should have API for defines outside of `<lang>_args`.

@xclaesse I think we can probably add that generic mechanism when people ask for it, but it's better to implement a feature with smaller scope first.

I would actually imagine this being implemented as:
```meson
executable('foo', 'foo.c',
           defines: {'static': ['FOO=bar', 'BAR=baz'],
                     'shared': ['BAH=blah']}
```

and then the defines get added to all possible languages. @jpakkane what do you think?

**Comment by jpakkane** [comment]
NO new kwargs only for defines. `-D` is standard, works with every compiler and is something people already know and use.

**Comment by xclaesse** [comment]
What about:
```meson
library('foo', 'foo.c',
  static_args : {'c' : ['-DFOO'],
                        'cpp': ['-DBAR']}
)
```

**Comment by xclaesse** [comment]
Or

```meson
library('foo', 'foo.c',
  c_args : {'static' : ['-DFOO'],
                 'shared': ['-DBAR']}
)
```

**Comment by NickeZ** [comment]
I like the last one. I think it would be cool if it was supported on:

```meson
cargs = {'shared_library' : ['-DH5_BUILT_AS_DYNAMIC_LIB']}
lib = library('lib', 'lib.c')
dep = declare_dependency(link_with: lib, compile_args: cargs)
executable('app', 'app.c', dependencies: dep) # <-- should get -DH5_BUILT_AS_DYNAMIC_LIB
```

Unless I completely missunderstood the `compile_args` kwarg. I haven't actually used it like this yet.

**Comment by nioncode** [comment]
For the method in https://github.com/mesonbuild/meson/issues/3304#issuecomment-407870518 to work with dllimport / dllexport on Windows, I think there must be a way to differentiate between defines that are used when the library gets compiled and ones that get used when it is used as a dependency.

**Comment by jpakkane** [comment]
Any flags specified in the `library` call are only used when the library is being compiled. Flags that should be used when consuming the library should be put in a `declare_dependency`.

**Comment by nioncode** [comment]
Ah true, then this would probably work quite nicely.

**Comment by tristan957** [comment]
@xclaesse you said earlier in this thread you had a patch for this around 2 years ago. Any thoughts on picking that up? Or I could potentially if your old work no longer applies/don't have time.

Seems like the dictionary version of `<language>_args` is preferred.

**Comment by xclaesse** [comment]
@tristan957 other developers did not like the dictionary approach. The last suggested idea was to do that with add_project_arguments(): https://github.com/mesonbuild/meson/pull/4159#issuecomment-782122257.

**Comment by dcbaker** [comment]
Let me add to this mess. Rust has several additional library (crate) types : c static, c dynamic, rust static, and rust dynamic. So we have at least two additional options here that I need today. So, I'd be in favor of the dictionary, but in a flat `List<str> | Dict<str, List<str>>` way. ie:
```meson
library(
  c_args : {
     'common': ['foo', 'bar'],
     'static' : '-DSTATIC',
     'shared' : '-DSHARED',
  },
  cpp_args : 'foo',
)
```
And for rust that could like like:
```meson
library(
  rust_args : {
    'common' : ['--cfg', 'always'],
    'rlib': ['--cfg', 'opt'],  # rust static
    'static': ['--cfg', 'static'],
    'cdylib': ['--cfg', 'dylib'],
  }
)
```

**Comment by dcbaker** [comment]
I'm at this point we need to agree to paint the dang shed, even if it's not perfect because this is starting to be a real problem.

**Comment by bruchar1** [comment]
I'd be in favor of `[<target_type>_]<lang>_args` (and `[<target_type>_]link_args`, like the initial proposal, as it keeps the syntax simpler, and it avoids the need of an artificial `all` or `common` target type.

I'm also in favor of having a target_type argument to `add_project_arguments` and `add_project_link_arguments`. I think both approaches are complementary.

**Comment by xclaesse** [comment]
> but in a flat `List<str> | Dict<str, List<str>>`

That's what I suggested 4 yearts ago too: https://github.com/mesonbuild/meson/issues/3304#issuecomment-407865753. Not sure we need common, as you can easily have a variable for that and do `{'static': common + []}`, but that's just a detail. IMHO we should just go ahead with that dict.

**Comment by xclaesse** [comment]
Let's just do it: https://github.com/mesonbuild/meson/pull/11501

**Comment by bruchar1** [comment]
Can this be closed, now that #11981 is merged?
